File: core/src/main/java/com/starshootercity/abilities/FallImmunity.java
package com.starshootercity.abilities;

import com.starshootercity.OriginSwapper;
import net.kyori.adventure.key.Key;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageEvent;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class FallImmunity implements VisibleAbility, Listener {
    @EventHandler
    public void onEntityDamage(EntityDamageEvent event) {
        if (event.getEntity() instanceof Player player) {
            AbilityRegister.runForAbility(player, getKey(), () -> {
                if (event.getCause() == EntityDamageEvent.DamageCause.FALL) {
                    event.setCancelled(true);
                }
            });
        }
    }

    @Override
    public @NotNull Key getKey() {
        return Key.key("origins:fall_immunity");
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getDescription() {
        return OriginSwapper.LineData.makeLineFor(
                    "You never take fall damage, no matter from which height you fall.",
                    OriginSwapper.LineData.LineComponent.LineType.DESCRIPTION);
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getTitle() {
        return OriginSwapper.LineData.makeLineFor(
                "Acrobatics",
                OriginSwapper.LineData.LineComponent.LineType.TITLE
        );
    }
}

File: core/src/main/java/com/starshootercity/abilities/FallImmunity.kt
package com.starshootercity.abilities

import com.starshootercity.OriginSwapper.LineData.Companion.makeLineFor
import com.starshootercity.OriginSwapper.LineData.LineComponent
import com.starshootercity.abilities.Ability.AbilityRunner
import net.kyori.adventure.key.Key
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageEvent

class FallImmunity : VisibleAbility, Listener {

    @EventHandler
    fun onEntityDamage(event: EntityDamageEvent) {
        if (event.cause != EntityDamageEvent.DamageCause.FALL) return

        runForAbility(event.entity, AbilityRunner { player ->
            event.isCancelled = true
        })
    }

    override fun getKey(): Key {
        return Key.key("origins:fall_immunity")
    }

    override fun getDescription(): MutableList<LineComponent?> {
        return makeLineFor(
            "You never take fall damage, no matter from which height you fall.",
            LineComponent.LineType.DESCRIPTION
        )
    }

    override fun getTitle(): MutableList<LineComponent?> {
        return makeLineFor(
            "Acrobatics",
            LineComponent.LineType.TITLE
        )
    }
}
